Adding get client feature

abhinavg4 opened the pull request:
## Description
This PR adds a get client function.

It sets up prometheus and graphana for the user and also allows customizing almost any parameter for Ray cluster.
## Usage
`get_client()`
```python
# Add snippet demonstrating usage
```
## Checklist
<!--
Note: All commits need to be signed and signed off. This can be done via `-sS` flags while commiting
`git commit -sS -m "...."
-->
- [ ] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [ ] New or Existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

abhinavg4 commented on ray-curator/ray_curator/core/client.py:
```diff
@@ -0,0 +1,365 @@
+    ray_dashboard_host: str = DEFAULT_RAY_DASHBOARD_HOST,
+    num_gpus: int | None = None,
+    num_cpus: int | None = None,
+    enable_object_spilling: bool = False,
```
We need to use ray start so there is no 1-1 mapping of the python and the command line arguments. I can use something like onto_slurm for making RayConfig and we can add stuff to it if required .

abhinavg4 commented on ray-curator/ray_curator/core/client.py:
```diff
@@ -0,0 +1,365 @@
+        return_dict["prometheus_web_port"] = prometheus_web_port
+        return_dict["grafana_web_port"] = grafana_web_port
+
+        downloaded, file_name = download_prometheus()
+        if not downloaded:
+            logger.error("Failed to download Prometheus.")
+            sys.exit(1)
```
Hmmm, we can have a parameter, but I think in the spirit of making things e2e, I would prefer we do download

abhinavg4 replied:
Added a param which is default to True. Open to making it False as default

abhinavg4 commented on ray-curator/ray_curator/core/client.py:
```diff
@@ -0,0 +1,365 @@
+        prometheus_dir = file_path.rstrip(".tar.gz")  # noqa: B005
+
+        # Write the prometheus.yml file to the NEMO_CURATOR_METRICS_PATH directory
+        prometheus_config_path = os.path.join(DEFAULT_NEMO_CURATOR_METRICS_PATH, "prometheus.yml")
+        with open(prometheus_config_path, "w") as f:
+            f.write(
+                PROMETHEUS_YAML_TEMPLATE.format(
+                    service_discovery_path=os.path.join(ray_temp_dir, "prom_metrics_service_discovery.json")
+                )
+            )
+
+        prometheus_cmd = [
+            f"{prometheus_dir}/prometheus",
+            "--config.file",
+            str(prometheus_config_path),
+            "--web.enable-lifecycle",
+            f"--web.listen-address=:{prometheus_web_port}",
```
100%

abhinavg4 commented on ray-curator/ray_curator/core/client.py:
```diff
@@ -0,0 +1,365 @@
+        # -----------------------------
+        try:
+            # Determine download URL based on architecture
+            arch = platform.machine()
+            if arch not in ("x86_64", "amd64"):
+                logger.warning(
+                    "Automatic Grafana installation is only tested on x86_64/amd64 architectures. "
+                    "Please install Grafana manually if the following steps fail."
+                )
```
Yes

praateekmahajan commented on ray-curator/tests/test_get_ray_client.py:
```diff
@@ -0,0 +1,106 @@
+
+    # TODO: This will kill all the ray processes. Find a better way to do this.
+    # Cleanup : stop any Ray processes we started
+    subprocess.run(["ray", "stop", "--force"], check=False, capture_output=True)  # noqa: S603, S607
```
yup this todo might be important.. would hate for it to kill my existing job if i end up running it on my cluster while i have the job.. i had to hack it in conftest  in this PR too https://github.com/NVIDIA-NeMo/Curator/pull/844

abhinavg4 replied:
Hmm, any good solution for this? In the case of getting a client, the ray init happens in a subprocess, so it's not straightforward to do a shutdown on it?

praateekmahajan reviewed (changes requested):
Took a high level look it mostly looks good.. here are some nits
1. `get_client` by default could only get client. and we grafana prometheus behind a boolean which is false.
2. let's refactor it such that grafana prometheus are behind functions so `get_client` is super trivial to follow.. at the end of the day get_client should get client and everything else is nice to have
3. instead of core move it to core/client; similar for test let's follow the same nested structure instead of parent level
4. I love your test for `test_two_clusters_metrics_visible`. If we can add similarly more tests that'll be great too!

abhinavg4 commented on ray-curator/ray_curator/core/client.py:
```diff
@@ -0,0 +1,365 @@
+    else:
+        ray_dashboard_port = get_next_free_port(ray_dashboard_port)
+        ray_metrics_port = get_next_free_port(ray_metrics_port)
+        ray_port = get_next_free_port(ray_port)
```
Can you please elaborate on this?

abhinavg4 commented:
@praateekmahajan 

1. I added a boolean, it's True by default right now but open to making it False.
2. Yeah, refactored it that way.
3. Hmmm, it's already inside a client file. Do you mean put the 2 files inside a folder? I was hoping that Constant.py and util.py will be same as we grow core more. As for test, yes let me nest it.
4. Any tests that come to mind?

sarahyurick commented on ray-curator/ray_curator/core/client.py:
```diff
@@ -0,0 +1,156 @@
+            return_dict["prometheus_web_port"] = prometheus_web_port
+            return_dict["grafana_web_port"] = grafana_web_port
+
+            downloaded, file_name = download_prometheus()
+            if not downloaded:
+                error_msg = "Failed to download Prometheus."
+                logger.error(error_msg)
+                raise RuntimeError(error_msg)
+
+            file_path = move_and_extract_prometheus(file_name)
+
+            prometheus_dir = file_path.rstrip(".tar.gz")  # noqa: B005
```
From my understanding, Prometheus and Grafana should not be strict requirements here... we should have more flexibility.

abhinavg4 replied:
Hmmm, I just wanted to make sure we are supporting end-to-end use case. 

> 
This function is a bit overloaded IMO.

I agree. Especially since this will get more overloaded overtime as we add support for multinode. I have added a flag to disable the Prometheus and Grafana setup. Can I make that False by default?  I can factor out the full Prometheus and Grafana setup login out and just add a function too, but would wanna keep the loginc e2e here.

sarahyurick replied:
I meant why is the current code path:
```
if is_prometheus_running() or is_grafana_running():
    # Skip setup, etc.
else:
   download_prometheus()
   # Further down
   download_grafana()
```
? Is there not a way we can verify if they are downloaded already, and if not print an error/warning with the appropriate instructions to download?

VibhuJawa commented on ray-curator/ray_curator/core/utils.py:
```diff
@@ -0,0 +1,134 @@
+        os.environ["RAY_MAX_LIMIT_FROM_DATA_SOURCE"] = str(API_LIMIT)
+        os.environ["XENNA_RAY_METRICS_PORT"] = str(ray_metrics_port)
+
+    ip_address = socket.gethostbyname(socket.gethostname())
+    ray_command = ["ray", "start", "--head"]
+    ray_command.extend(["--node-ip-address", ip_address])
+    ray_command.extend(["--port", str(ray_port)])
+    ray_command.extend(["--metrics-export-port", str(ray_metrics_port)])
+    ray_command.extend(["--dashboard-host", ray_dashboard_host])
+    ray_command.extend(["--dashboard-port", str(ray_dashboard_port)])
+    ray_command.extend(["--temp-dir", ray_temp_dir])
+    ray_command.extend(["--disable-usage-stats"])
+    if enable_object_spilling:
+        ray_command.extend(
+            [
+                "--system-config",
+                '{"local_fs_capacity_threshold": 0.95, "object_spilling_config": "{ "type": "filesystem", "params": {"directory_path": "/tmp/ray_spill", "buffer_size": 1000000 } }"}',
```
Can we file and link an issue to ensure  to ray , we can work with them on adding the missing features on ray.init please.

VibhuJawa commented on ray-curator/ray_curator/core/utils.py:
```diff
@@ -0,0 +1,378 @@
+    return logger
+
+
+def init_or_connect_to_cluster(  # noqa: PLR0913
```
Could you point me to where we  connect to the cluster here, seems like we only start it using CLI ?

sarahyurick commented on ray-curator/ray_curator/core/client.py:
```diff
@@ -0,0 +1,144 @@
+        num_cpus: The number of CPUs to use.
+        enable_object_spilling: Whether to enable object spilling.
+
+    Note:
+        To start monitoring services (Prometheus and Grafana), use the standalone
+        start_prometheus_grafana.py script separately.
```
Does this script exist yet?

sarahyurick commented on ray-curator/ray_curator/core/client.py:
```diff
@@ -0,0 +1,144 @@
+        if self.ray_process:
+            self.ray_process.kill()
+            self.ray_process.wait()
+            # Current there is no good way of stopping a particular ray cluster. https://github.com/ray-project/ray/issues/54989
+            # We kill the ray gcs process to stop the cluster. but still we have some ray processes running.
+            msg = "Nemo Curator has stopped the ray cluster it started by killing the ray gcs process. "
+            msg += "It is advised to wait for a few seconds before running any ray commands to ensure ray can clenup other processes."
+            msg += "If you are seeing any ray commands like `ray status` failing, please ensure /tmp/ray/ray_current_cluster has correct information."
```
```suggestion
            # Currently there is no good way of stopping a particular Ray cluster. https://github.com/ray-project/ray/issues/54989
            # We kill the Ray GCS process to stop the cluster, but still we have some Ray processes running.
            msg = "NeMo Curator has stopped the Ray cluster it started by killing the Ray GCS process. "
            msg += "It is advised to wait for a few seconds before running any Ray commands to ensure Ray can cleanup other processes."
            msg += "If you are seeing any Ray commands like `ray status` failing, please ensure /tmp/ray/ray_current_cluster has correct information."
```

VibhuJawa commented on ray-curator/ray_curator/core/utils.py:
```diff
@@ -0,0 +1,119 @@
+    num_gpus: int | None = None,
+    num_cpus: int | None = None,
+    enable_object_spilling: bool = False,
+    block: bool = True,
```
I dont think we should support non blocking mode today, lets remove this option for now please as if non blocking we can not kill it

abhinavg4 replied:
I think once ray allows to stop a sepcific server this might be used. I can add an assert that block is always true.

abhinavg4 replied:
Ok we are not passing `block` as a param from get_client class (which means it's not an option). But incase the user wants to use this util without `block`, I don't see a reason why they should not be able to ?

VibhuJawa commented on ray-curator/tests/core/test_get_ray_client.py:
```diff
@@ -0,0 +1,63 @@
+        with open(os.path.join(ray_tmp, "ray_current_cluster")) as f:
+            assert f.read() == f"127.0.0.1:{client.ray_port}"
+        client.stop()
+
```
Can you add a asset here to verify stop actually worked ?

abhinavg4 replied:
Hmm there is no easy way to verify that the stop actually worked.

VibhuJawa reviewed (changes requested):
Mostly looks good, requested some minor changes

abhinavg4 commented on ray-curator/ray_curator/core/client.py:
```diff
@@ -0,0 +1,144 @@
+            self.ray_process.kill()
+            self.ray_process.wait()
+            # Current there is no good way of stopping a particular ray cluster. https://github.com/ray-project/ray/issues/54989
+            # We kill the ray gcs process to stop the cluster. but still we have some ray processes running.
```
HI yes, if you wait they should eventually get killed.

abhinavg4 replied:
I'll make a ticket of this.

VibhuJawa reviewed (approved):
LGTM. Thanks for all the changes

abhinavg4 commented:
/ok to test [2bdd0b0](https://github.com/NVIDIA-NeMo/Curator/pull/834/commits/2bdd0b04ecd7980bb4b7a350b0c49d00a6a6c067)
